src/policy/ancestor_packages.cpp

```diff
@@ -0,0 +1,260 @@
+    // Now populate the descendant caches
+    for (const auto& [txid, entry] : m_txid_to_entry) {
+        for (const auto& anc_txid : entry.ancestor_subset) {
+            m_txid_to_entry.at(anc_txid).descendant_subset.insert(txid);
+        }
+    }
+}
+
+Package AncestorPackage::Txns() const
+{
+    Package result;
+    std::transform(m_txns.cbegin(), m_txns.cend(), std::back_inserter(result),
+                   [](const auto refentry){ return refentry.get().tx; });
+    return result;
+}
+
+Package AncestorPackage::FilteredTxns() const
+{
+    Package result;
+    for (const auto entryref : m_txns) {
+        if (entryref.get().m_state == PackageEntry::State::PACKAGE) result.emplace_back(entryref.get().tx);
+    }
+    return result;
+}
+
+std::optional<std::vector<CTransactionRef>> AncestorPackage::FilteredAncestorSet(const CTransactionRef& tx) const
+{
+    const auto& entry_it = m_txid_to_entry.find(tx->GetHash());
+    if (entry_it == m_txid_to_entry.end()) return std::nullopt;
+    const auto& entry = entry_it->second;
+    if (entry.m_state == PackageEntry::State::DANGLING) return std::nullopt;
+
+    std::vector<CTransactionRef> result;
+    result.reserve(entry.ancestor_subset.size());
+    for (const auto entryref : m_txns) {
+        if (entry.ancestor_subset.count(entryref.get().tx->GetHash()) > 0) {
+            // All non-PACKAGE ancestor_subset transactions should have been removed when the Skip* happened.
+            if (Assume(entryref.get().m_state == PackageEntry::State::PACKAGE)) {
+                result.emplace_back(entryref.get().tx);
+            }
+        }
+    }
+    return result;
+}
+
+std::optional<std::pair<CAmount, int64_t>> AncestorPackage::GetFeeAndVsize(const CTransactionRef& tx) const
+{
+    const auto& entry_it = m_txid_to_entry.find(tx->GetHash());
+    if (entry_it == m_txid_to_entry.end()) return std::nullopt;
+
+    const auto& entry = entry_it->second;
+    if (entry.m_state == PackageEntry::State::DANGLING || !entry.fee.has_value() || !entry.vsize.has_value()) return std::nullopt;
+    return std::make_pair(entry.fee.value(), entry.vsize.value());
+}
+
+std::optional<std::pair<CAmount, int64_t>> AncestorPackage::FilteredAncestorFeeAndVsize(const CTransactionRef& tx) const
+{
+    const auto& entry_it = m_txid_to_entry.find(tx->GetHash());
+    if (entry_it == m_txid_to_entry.end()) return std::nullopt;
+    const auto& entry = entry_it->second;
+    if (entry.m_state == PackageEntry::State::DANGLING || !entry.fee.has_value() || !entry.vsize.has_value()) return std::nullopt;
+
+    CAmount total_fee{0};
+    int64_t total_vsize{0};
+    for (const auto& txid : entry.ancestor_subset) {
+        const auto& anc_entry = m_txid_to_entry.at(txid);
+        // All non-PACKAGE ancestor_subset transactions should have been removed when the Skip* happened.
+        if (Assume(anc_entry.m_state == PackageEntry::State::PACKAGE)) {
```

Comment:
89913ee14e8e2a3247191966c66b3c3fcbeb45b8

Is it safe to still return a non-null result if this fails?

Reply:
Yes, this is just asserting that we cleaned up in the skipping functions the way we were supposed to. I think as long as this is ancestor set-shaped there isn't much danger.